Stabilize const_cstr_from_ptr (CStr::from_ptr, CStr::count_bytes)

dtolnay opened the pull request:
Completed the pair of FCPs https://github.com/rust-lang/rust/issues/113219#issuecomment-2016939401 + https://github.com/rust-lang/rust/issues/114441#issuecomment-2016942566.

`CStr::from_ptr` is covered by just the first FCP on its own. `CStr::count_bytes` requires the approval of both FCPs. The second paragraph of the first link and the last paragraph of the second link explain the relationship between the two FCPs. As both have been approved, we can proceed with stabilizing `const` on both of these already-stable functions.

dtolnay committed:
Stabilize const_cstr_from_ptr (CStr::from_ptr, CStr::count_bytes)

tgross35 commented on library/core/src/ffi/c_str.rs:
```diff
@@ -742,6 +740,8 @@ impl AsRef<CStr> for CStr {
 /// The pointer must point to a valid buffer that contains a NUL terminator. The NUL must be
 /// located within `isize::MAX` from `ptr`.
 #[inline]
+#[rustc_const_stable(feature = "const_cstr_from_ptr", since = "CURRENT_RUSTC_VERSION")]
```
Assuming the attribute on a nonpublic function allows using `const_strlen` everywhere without adding `const_eval_select` each time?

workingjubilee replied:
I would prefer not to add such misuses of stability attributes. It makes it harder to review internal changes, because an important part of T-libs reviews when you _aren't_ reviewing an addition API is making sure that you are not, in fact, adding API. This weakens such a defense and makes it easier for a reviewer to get confused.

tgross35 replied:
In what way is this a misuse? The lang and libs teams FCPd having a `const_strlen` available in some form for internal use, currently via `const_eval_select` but that is an implementation detail. Marking it as "stable" (but of course nonpublic) seems to accurately reflect that status, as opposed to leaking the `const_eval_select` implementation detail around everywhere it is used.

dtolnay replied:
Right @tgross35, I believe this is const stability working as intended. The attribute enforces that future changes to this function do not introduce any unapproved implementation, because this private function now participates in the implementation of some public stable const function and the use of any new weird const internals here would become a stable commitment. Flagging the presence of a stable commitment is the point of a stability attribute.

@RalfJung endorsed the same thing recently in https://github.com/rust-lang/rust/pull/124941#discussion_r1664676739.

tgross35 replied:
It would probably be okay to rename `const_strlen` to just `strlen` at some point to make it more clear that's just our implementation of `strlen` now (avoiding confusion like https://github.com/rust-lang/rust/pull/127444).

workingjubilee replied:
I believe it should have a comment disclaiming its status, then, rather than passing unremarked, because this strikes me as a nonintuitive combination of attributes when placed in a public module.

Or it could be marked `#[unstable]`, as the other function cited is.

dtolnay replied:
I have added `unstable` in https://github.com/rust-lang/rust/pull/127433/commits/7f1518bddd76ad8e3a743278fd983c35d11b2411.

workingjubilee replied:
Thank you. "Fix this one attribute" or "oops this stabilization missed a spot" or "make this thing public that obviously was intended to be public... right?" are all fairly common PRs.

dtolnay committed:
Add instability attribute on private const_strlen function

A `rustc_const_stable` attribute by itself has nonintuitive purpose when
placed in a public module.

Separately, it would probably be okay to rename `const_strlen` to just
`strlen` to make it more clear this is our general-purpose
implementation of strlen now, not something specifically for const
(avoiding confusion like in PR 127444).

workingjubilee commented:
@bors r+

bors commented:
:pushpin: Commit 7f1518bddd76ad8e3a743278fd983c35d11b2411 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 7f1518bddd76ad8e3a743278fd983c35d11b2411 -->
<!-- homu: {"type":"Approved","sha":"7f1518bddd76ad8e3a743278fd983c35d11b2411","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
